### PR TITLE
[Merged by Bors] - Return eth1-related data via the API

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

- Related to #1691

## Proposed Changes

Adds the following API endpoints:

- `GET lighthouse/eth1/syncing`: status about how synced we are with Eth1.
- `GET lighthouse/eth1/block_cache`: all locally cached eth1 blocks.
- `GET lighthouse/eth1/deposit_cache`: all locally cached eth1 deposits.

Additionally:

- Moves some types from the `beacon_node/eth1` to the `common/eth2` crate, so they can be used in the API without duplication.
- Allow `update_deposit_cache` and `update_block_cache` to take an optional head block number to avoid duplicate requests.

## Additional Info

TBC

### CODE REVIEW

**Comment by paulhauner** [comment]
Hey @pawanjay176 would you be interested in reviewing this? :)

File: beacon_node/eth1/src/service.rs [hunk 1]
```diff
@@ -301,9 +316,22 @@ impl Service {
     pub async fn update(
         &self,
     ) -> Result<(DepositCacheUpdateOutcome, BlockCacheUpdateOutcome), String> {
+        let endpoint = &self.config().endpoint.clone();
```
Comment (pawanjay176) [inline beacon_node/eth1/src/service.rs:319]: We could save a call to the eth1 node by using `eth_getBlockByNumber` with 'latest' in the params to get the block corresponding to remote head.

File: book/src/api-lighthouse.md [hunk 2]
```diff
@@ -178,6 +178,158 @@ See [Validator Inclusion APIs](./validator-inclusion.md).
...
+- `eth1_node_sync_status_percentage` (float): An estimate of how far the head of the
+	Eth1 node is from the head of the Eth1 chain.
+	- `100.0` indicates a fully synced Eth1 node.
+	- `0.0` indicates an Eth1 node that has not verified any blocks past the
+		genesis block.
+- `lighthouse_is_cached_and_ready`: Is set to `true` if the caches in the
+	beacon node are ready for block production.
+	- This value might be set to
+	`false` whilst `eth1_node_sync_status_percentage == 100.0` if the beacon
+	node is still build its internal cache.
```
Comment (pawanjay176) [inline book/src/api-lighthouse.md:207]:
```suggestion
	node is still building its internal cache.
```

File: book/src/api-lighthouse.md [hunk 3]
```diff
@@ -178,6 +178,158 @@ See [Validator Inclusion APIs](./validator-inclusion.md).
...
+      "timestamp": 1603173353,
+      "number": 3606742,
+      "deposit_root": "0xd24920d936e8fb9b67e93fd126ce1d9e14058b6d82dcf7d35aea46879fae6dee",
+      "deposit_count": 88911
+    }
+  ]
+}
+```
+
+### `/lighthouse/eth1/block_cache`
```
Comment (pawanjay176) [inline book/src/api-lighthouse.md:263]: `/lighthouse/eth1/deposit_cache` is already present above. This is a repetition

File: beacon_node/client/src/builder.rs [hunk 4]
```diff
@@ -590,7 +591,7 @@ where
                 })?
         };
 
-        self.eth1_service = None;
+        self.eth1_service = Some(backend.core.clone());
```
Comment (pawanjay176) [inline beacon_node/client/src/builder.rs:594]: Any idea why we set this to `None` earlier? curious again :)

**Review by pawanjay176 — APPROVED** [review]
Looks good. Just a few minor suggestions and questions :)

Apart from this, I had one minor peeve with the `update_deposit_cache` for a long time. Currently, the `max_log_requests_per_update` config param in eth1 service is always set to `u64::max()`. Because of this, when we start building the eth1 cache from scratch, the call to `update_deposit_cache` only returns after getting all logs upto current head. If even a single query to the eth1 node fails, we have to start over building the deposit cache.

I think it'd be good if we set `max_log_requests_per_update` to some smaller value so that we don't try to get all the logs with a single `update_deposit_cache` call.

File: beacon_node/eth1/src/service.rs [hunk 5]
```diff
@@ -301,9 +316,22 @@ impl Service {
...
+        let remote_head_block_number =
```
Comment (pawanjay176) [inline beacon_node/eth1/src/service.rs:320]: We could save a call to the eth1 node by using eth_getBlockByNumber with 'latest' in the params to get the block corresponding to remote head.

Comment (paulhauner) [reply beacon_node/eth1/src/service.rs:320]: Nice one, I'll do this!

Comment (paulhauner) [reply beacon_node/eth1/src/service.rs]: Done here: https://github.com/sigp/lighthouse/pull/1797/commits/b0d037375eeab42d05e1a5653da0e56ac276ce60

:)

File: beacon_node/client/src/builder.rs (hunk 4 above)
Comment (paulhauner) [reply beacon_node/client/src/builder.rs:594]: Hmm.. no I can't :sweat_smile:

File: beacon_node/beacon_chain/src/eth1_chain.rs [hunk 6]
```diff
@@ -19,6 +21,9 @@ use types::{
 type BlockNumber = u64;
 type Eth1DataVoteCount = HashMap<(Eth1Data, BlockNumber), u64>;
 
+/// We will declare ourself synced with the Eth1 chain, even if we are this many blocks behind.
+const ETH1_SYNC_TOLERANCE: u64 = 8;
```
[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/eth1_chain.rs:27]: Nope, no reason. I added a note to that measure :)

[the remark this answers is not in the document]
Comment (paulhauner) [reply beacon_node/beacon_chain/src/eth1_chain.rs:27]: Well, 8 just felt like a decent number to me. I can't back it up, though :sweat_smile:

**Comment by paulhauner** [comment]
All comments addressed!

**Review by pawanjay176 — APPROVED** [review]

**Pull request closed without merging**